chainermn/communicators/_base.py

```diff
@@ -1,351 +1,87 @@
-            self.mpi_comm.Bcast(buf, root)
-            return x
-        else:
-            msgtype = None
-            msgtype = self.mpi_comm.bcast(msgtype, root)
-            shape = msgtype.shapes[0]
-            buf = numpy.empty(numpy.prod(shape), dtype=numpy.float32)
-            self.mpi_comm.Bcast(buf, root)
-            return buf.reshape(shape)
+        raise NotImplementedError()
 
-    def gather(self, x, root=0):
-        """A primitive of inter-process gather communication.
+    # on ndarrays and such
+    def send(self, data, dest, tag):
+        raise NotImplementedError()
 
-        This method tries to invoke gather communication within the
-        communicator. All processes in the communicator are expected to
-        invoke ``gather()``. This method relies on mpi4py fast communication
-        optimized for numpy arrays, as well as ``send()`` and ``recv()``.
+    def recv(self, source, tag):
+        raise NotImplementedError()
 
-        Note that this method can only handle the same shapes of data
-        over all processes, and cannot handle tuple data.
+    def bcast(self, data, max_buf_len=None, root=0):
+        raise NotImplementedError()
 
-        Args:
-            x (numpy.array): Array to be gathered.
+    def gather(self, data, root=0):
+        raise NotImplementedError()
 
-        Returns:
-            ys (numpy.ndarray):
-                Received arrays with shape (#proc, [data-shape]).
-                ``None`` for non-root processes.
-        """
-        chainer.utils.experimental(
-            'chainermn.communicators.CommunicatorBase.gather')
+    def allreduce(self, data):
+        raise NotImplementedError()
 
-        is_master = self.mpi_comm.rank == root
+    # on objects
+    def send_obj(self, obj, dest, tag):
+        raise NotImplementedError()
 
-        msgtype = _MessageType(x)
-        msgtypes = self.mpi_comm.gather(msgtype, root)
+    def recv_obj(self, source, tag):
+        raise NotImplementedError()
 
-        # Type check.
-        if is_master:
-            shape = msgtype.shapes[0]
-            for msgtype in msgtypes:
-                if msgtype.is_tuple:
-                    raise TypeError('gather cannot handle tuple data')
+    def bcast_obj(self, data, max_buf_len=None, root=0):
+        raise NotImplementedError()
 
-                assert len(msgtype.shapes) == 1
+    def gather_obj(self, data, root=0):
+        raise NotImplementedError()
 
-                if msgtype.shapes[0] != shape:
-                    raise ValueError(
-                        'gather cannot handle different shapes of data')
+    def allreduce_obj(self, model, op=None):
+        raise NotImplementedError()
 
-        if x.dtype != numpy.float32:
-            raise ValueError('gather only support dtype == numpy.float32')
+    # Special communication method on grads and data of models
+    def bcast_data(self, model, max_buf_len=None, root=0):
```

Comment:
`bcast_data` and `broadcast_data` are for our internal use (from multi node optimizers), and, at least for now, we don't need `max_buf_len` and `root`. I propose to keep it simpler by removing unused arguments.